### DISCUSSION TITLE
Collaboration & Other Use Cases

### DISCUSSION THREAD

**Discussion by Stuckya** [rednblkx/HomeKey-ESP32#4] [General]
I've been working on a similar project. Albeit a bit more focused on the hardware side at the moment. Designing a weather resistant (_something akin to IP65_) NFC reader that could be flashed with custom firmware, like the one in this repository.
The hardware design is based around POE (or 24v, whichever is more readily available), an ESP32, and the Elechouse PN532 module.

Instead of duplicating efforts, I'd rather combine efforts and make contributions to an existing public project.

The current structure of this project implies a fairly specific use-case at the moment. (_A single physical locking mechanism controlled by the ESP32 where the NFC module is connected to that same ESP32. Additionally, the Home Key is provisioned to that specific ESP32 device._)

I'd love to contribute to the project, but desire some additionally flexibility to compose some other use-cases. (_I see an MQTT ticket on the project board, but this still might not be flexible enough for other use-cases I have in mind._)

Let me know what you think about these use-cases:
- NFC support for other authentication methods, such as MIFARE DESFire, in addition Home Key. Maybe this becomes a protocol library we can compose into another package that has a web-interface and composes additional NFC protocols?
- Configuring multiple ESP32 "readers" using a shared Home Key. (Using the same key for front door, back door, garage side-door, garage door, gate, etc.)
    - Remote authorization. With a shared key, let the reader handle the attestation flow of the Home Key credential, but remotely validate if that key can access a specific door / reader. (_Note: MQTT might work for this use-case._)
    - Remote locking mechanism. For example, a hardwired door strike in a different location from the reader or z-wave device control. (_Note: MQTT might work for this use-case._)

**Comment by rednblkx** [comment]
Hello Andrew,

Thank you for taking an interest in this project!

To answer you inquiry shortly, i am envisioning this repository to become eventually a library that is part of a much greater collection of authentication systems package powered by an ESP microcontroller.

For the moment, however, I'm concentrating on making the code more stable and more or less efficient. As you will see in #3 there are some problems.

Regarding your suggested use cases, as mentioned above, I'm open to implementing more protocols and extending the usability of the project once I can say that the HomeKey implementation is ready.

Additionally, my goal is to make this a device that you can put anywhere and trigger anything through MQTT and such and you will have Remote locking mechanism and while you could hardwire the ESP to a lock and trigger it like that, i'm not planning on adding support for that(though not set in stone).

Remote authorisation is on the map, as I'll be exposing the issuerId, which is unique per iCloud acc, and the endpointId, which is unique per device via MQTT, so you can make decisions based on a person or a device.

Having that said, you are more than welcome to contribute to this project if you wish.
